Maple2.File.IO/Nif/NifDocument.cs

```diff
@@ -0,0 +1,247 @@
+        ReadingBlock = block;
+    }
+
+    public void Parse() {
+        try {
+            ParseImplementation();
+        } catch (Exception ex) {
+            StringBuilder builder = new StringBuilder();
+
+            builder.Append($"Error reading nif: [{RelPath}]");
+
+            if (VersionString.Length > 0) {
+                builder.Append($" ({VersionString})");
+            }
+
+            if (ReadingBlock is not null) {
+                builder.AppendLine($" in block [{ReadingBlock.BlockIndex}] {ReadingBlock.BlockType} \"{ReadingBlock.Name}\"");
+            }
+
+            throw new Exception(builder.ToString(), ex);
+        }
+    }
+
+    private void ParseImplementation() {
+        int index = ReadHeaderString();
+
+        if (header.Version[0] < 30) {
+            throw new NifVersionNotSupportedException($"[{RelPath}]: NIF version number too low: {header.HeaderString}. Parser is built for v30+");
+        }
+
+        index += 4;
+
+        bool isLittleEndian = fileData[index] != 0;
+
+        MemoryStream stream = new MemoryStream(fileData);
+        Reader = new EndianReader(stream, isLittleEndian != BitConverter.IsLittleEndian);
+
+        Reader.Advance(index + 1);
+
+        uint userVersion = Reader.ReadAdjustedUInt32();
+        int numBlocks = Reader.ReadAdjustedInt32();
+        int metaBlockSize = Reader.ReadAdjustedInt32();
+
+        Reader.Advance(metaBlockSize);
+
+        ushort numBlockTypes = Reader.ReadAdjustedUInt16();
+
+        if (numBlockTypes == 0) {
+            return;
+        }
+
+        header.BlockTypes = new string[numBlockTypes];
+        header.BlockTypeIndices = new ushort[numBlocks];
+        header.BlockSizes = new int[numBlocks];
+        blocks = new NifBlock?[numBlocks];
+
+        for (uint i = 0; i < numBlockTypes; i++) {
+            header.BlockTypes[i] = Reader.ReadAdjustedStringLen32();
+        }
+
+        for (uint i = 0; i < numBlocks; i++) {
+            header.BlockTypeIndices[i] = (ushort) (0x7FFFU & Reader.ReadAdjustedUInt16());
+        }
+
+        for (uint i = 0; i < numBlocks; i++) {
+            header.BlockSizes[i] = Reader.ReadAdjustedInt32();
+        }
+
+        uint numStrings = Reader.ReadAdjustedUInt32();
+        uint maxStringLength = Reader.ReadAdjustedUInt32();
+
+        header.Strings = new string[numStrings];
+
+        for (uint i = 0; i < numStrings; i++) {
+            header.Strings[i] = Reader.ReadAdjustedStringLen32();
+        }
+
+        uint numGroups = Reader.ReadAdjustedUInt32();
+
+        if (numGroups > 0) {
+            throw new NotSupportedException("NIF groups aren't currently supported");
+        }
+
+        for (int i = 0; i < numBlocks; ++i) {
+            string blockType = header.BlockTypes[header.BlockTypeIndices[i]];
+            int blockSize = header.BlockSizes[i];
+
+            long blockStart = Reader.BaseStream.Position;
+
+
+            NifBlock? block = GetBlock(i);
+
+            block?.Parse(this);
```

Comment:
do we not want to handle null here instead of silently failing?

Reply:
Block there should never be null since I is guaranteed within bounds. I originally wrote that bit before I made it default construct NifBlock objects for unimplemented nodes, so I should probably change the code to account for that & already having done null bounds checks.